### PR TITLE
Faster Accuracy metric

### LINKED ISSUE #2722: Slow accuracy metric
## 🐛 Bug

Accuracy class from metrics.classification is too slow, compared to pytorch vanilla implementation.

### To Reproduce

Try the following demo code.

#### Code sample

```python3
from time import time
import torch
from pytorch_lightning.metrics.classification import Accuracy


def test_acc(pred, targ, get_acc, steps=1000, rep=10):
    tsum, mean_res = 0, []
    for i in range(rep):
        res = []
        ts = time()
        for j in range(steps):
            res.append(get_acc(pred, targ))
        te = time()
        tsum += te - ts
        mean_acc = torch.stack(res).mean().item()
        mean_res.append(mean_acc)
    
    return (tsum / rep), mean_res


def demo(batch_size=32, num_classes=10, device='cuda', steps=100, rep=5):
    print("Testing  bs: %d,  num_cls: %d,  device: %s,  steps: %d,  rep: %d" % (batch_size, num_classes, device, steps, rep))

    pred = torch.randint(0, num_classes, (batch_size, )).to(device=device)
    targ = torch.randint(0, num_classes, (batch_size, )).to(device=device)

    pl_acc = Accuracy(num_classes=num_classes)
    def pt_acc(pred, targ):
        return pred.eq(targ).to(dtype=torch.float).mean()

    pl_time, pl_res = test_acc(pred, targ, pl_acc, steps, rep)
    pt_time, pt_res = test_acc(pred, targ, pt_acc, steps, rep)

    if pl_res != pt_res:
        print("Results mismatch!")
    print("Lightning: %2.6fs" % pl_time)
    print("Vanilla  : %2.6fs" % pt_time)
    print("")

if __name__ == "__main__":
    demo(32, 10)
    demo(512, 10)
    demo(32, 100)
    demo(512, 100)
    demo(32, 1000)
    demo(512, 1000)

    demo(32, 10, 'cpu')
    demo(512, 10, 'cpu')
    demo(32, 100, 'cpu')
    demo(512, 100, 'cpu')
    demo(32, 1000, 'cpu')
    demo(512, 1000, 'cpu')
```

On my machine, the results are like the following.

```
Testing  bs: 32,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.378017s
Vanilla  : 0.002766s

Testing  bs: 512,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.372391s
Vanilla  : 0.002776s

Testing  bs: 32,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 3.346141s
Vanilla  : 0.002812s

Testing  bs: 512,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 3.335049s
Vanilla  : 0.002660s

Testing  bs: 32,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 32.751692s
Vanilla  : 0.002666s

Testing  bs: 512,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 32.967594s
Vanilla  : 0.002762s

Testing  bs: 32,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.128131s
Vanilla  : 0.001317s

Testing  bs: 512,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.128481s
Vanilla  : 0.001403s

Testing  bs: 32,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 1.117403s
Vanilla  : 0.001275s

Testing  bs: 512,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 1.166188s
Vanilla  : 0.001330s

Testing  bs: 32,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 10.927701s
Vanilla  : 0.001229s

Testing  bs: 512,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 11.560013s
Vanilla  : 0.001310s
```

100x ~ 10000x more time is needed for pl's accuracy, depending on the num_classes argument.

### Expected behavior

Pytorch-lightning's accuracy function should take similar time compared to the pytorch vanilla implementation, and it should not depend on num_classes too much.

### Environment

```
* CUDA:
	- GPU:
		- GeForce GTX 1050 Ti
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.0
	- pyTorch_debug:     False
	- pyTorch_version:   1.5.1
	- pytorch-lightning: 0.8.1
	- tensorboard:       2.2.2
	- tqdm:              4.47.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.6.9
	- version:           #46~18.04.1-Ubuntu SMP Fri Jul 10 07:21:24 UTC 2020
```

I tested the same code from two other machines (each with 2080 ti and with V100), and they give similar results.

### Additional context

I'm not sure this issue falls into the bug report or the feature request category. Please relocate this as you like. :)

### PR DESCRIPTION
**Pull request by Diuven**
## What does this PR do?

Speeding up the Accuracy metric for classification tasks.

Fixes #2722 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
@SkafteNicki @Borda @PyTorchLightning/core-contributors I think we now have to decide if we want to use the given num_classes or the calculated ones (especially if they don't match). Opinions?

**Comment by Diuven** [comment]
> @SkafteNicki @Borda @PyTorchLightning/core-contributors I think we now have to decide if we want to use the given num_classes or the calculated ones (especially if they don't match). Opinions?

Actually, I updated the code to do the same thing with the previous code when it gets different `num_classes` from the real number of classes. So if you want to put off the decision or want to keep the behaviour, you can do so.

I think this behaviour (using the results only in the range specified by the `num_classes`) can be useful when the user want the stats of the certain classes.

But I do think that this is quite counter-intuitive, and removing it will make the function more clear and faster. We can just remove this behaviour for now and add it later on, with well designed structure.

This depends on the core developers' opinions, so I'm willing to update the PR correspondingly. Please share your thoughts! :smiley:

File: pytorch_lightning/metrics/functional/classification.py [hunk 1]
```diff
@@ -173,16 +173,56 @@ def stat_scores_multiple_classes(
...
+        tns = pred.size(0) - (tps + fps + fns)
+        sups.scatter_add_(0, target, torch.ones_like(match_true))
+
+        tps = tps[:num_classes]
+        fps = fps[:num_classes]
+        tns = tns[:num_classes]
+        fns = fns[:num_classes]
+        sups = sups[:num_classes]
+
+    if reduction == 'sum' or reduction == 'elementwise_mean':
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:210]:
```suggestion
    elif reduction == 'sum' or reduction == 'elementwise_mean':
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 2]
```diff
@@ -138,10 +138,10 @@ def stat_scores_multiple_classes(
         target: torch.Tensor,
         num_classes: Optional[int] = None,
         argmax_dim: int = 1,
+        reduction: str = 'none'
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:141]:
```suggestion
        reduction: str = 'none',
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 3]
```diff
@@ -173,16 +173,56 @@ def stat_scores_multiple_classes(
...
+    num_classes = get_num_classes(pred=pred, target=target, num_classes=num_classes)
 
-    tps = torch.zeros((num_classes,), device=pred.device)
-    fps = torch.zeros((num_classes,), device=pred.device)
-    tns = torch.zeros((num_classes,), device=pred.device)
-    fns = torch.zeros((num_classes,), device=pred.device)
-    sups = torch.zeros((num_classes,), device=pred.device)
-    for c in range(num_classes):
-        tps[c], fps[c], tns[c], fns[c], sups[c] = stat_scores(pred=pred, target=target, class_index=c)
+    pred = pred.view((-1, )).long()
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:178]: is the long (int64) really needed here?

Comment (Diuven) [reply pytorch_lightning/metrics/functional/classification.py:178]: Yes. `torch.scatter_add` function requires the 2nd argument (denoting 'index') to be `LongTensor`. If you would like, we can push that type conversion into the `if 'none'` block as the other parts doesn't require it.

Comment (justusschock) [reply pytorch_lightning/metrics/functional/classification.py:178]: @Diuven please do that, since this may otherwise also take some time even if not needed

Comment (Diuven) [reply pytorch_lightning/metrics/functional/classification.py]: OK. Updated accordingly.

**Comment by Diuven** [comment]
I'm keep getting some pip install errors on CI test intermittently. Can someone check if some code of mine are causing such problems?

**Comment by justusschock** [comment]
@Diuven This keeps happening from time to time, although we don't know why. Usually retriggering helps (what I did now) :)

File: pytorch_lightning/metrics/functional/classification.py [hunk 4]
```diff
@@ -138,10 +138,10 @@ def stat_scores_multiple_classes(
         target: torch.Tensor,
         num_classes: Optional[int] = None,
         argmax_dim: int = 1,
+        reduction: str = 'none',
```
Comment (justusschock) [inline pytorch_lightning/metrics/functional/classification.py:141]: Can you add a line in docstrings for this?

Comment (Diuven) [reply pytorch_lightning/metrics/functional/classification.py:141]: Oh, sure. I missed that.
Do we need updated document (which is on the readthedocs) about this also?

Comment (justusschock) [reply pytorch_lightning/metrics/functional/classification.py:141]: no, this is automatically parsed
